server/src/main/java/io/druid/server/coordinator/rules/LoadRule.java

```diff
@@ -50,187 +51,267 @@
+      return Objects::nonNull;
+    } else {
+      return s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
+    }
+  }
 
-      final MinMaxPriorityQueue<ServerHolder> serverQueue = params.getDruidCluster().getHistoricalsByTier(tier);
+  private static List<ServerHolder> getFilteredHolders(
+      final String tier,
+      final DruidCluster druidCluster,
+      final Predicate<ServerHolder> predicate
+  )
+  {
+    final NavigableSet<ServerHolder> queue = druidCluster.getHistoricalsByTier(tier);
+    if (queue == null) {
+      log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
+      return Collections.emptyList();
+    }
 
-      if (serverQueue == null) {
-        log.makeAlert("Tier[%s] has no servers! Check your cluster configuration!", tier).emit();
+    return queue.stream().filter(predicate).collect(Collectors.toList());
+  }
+
+  @Nullable
+  private static ServerHolder assignPrimary(
+      final Object2IntMap<String> targetReplicants,
+      final DruidCoordinatorRuntimeParams params,
+      final DataSegment segment
+  )
+  {
+    ServerHolder topCandidate = null;
+    for (final Object2IntMap.Entry<String> entry : targetReplicants.object2IntEntrySet()) {
+      final int targetReplicantsInTier = entry.getIntValue();
+      if (targetReplicantsInTier <= 0) {
         continue;
       }
 
-      final int maxSegmentsInNodeLoadingQueue = params.getCoordinatorDynamicConfig()
-                                                      .getMaxSegmentsInNodeLoadingQueue();
+      final String tier = entry.getKey();
 
-      Predicate<ServerHolder> serverHolderPredicate;
-      if (maxSegmentsInNodeLoadingQueue > 0) {
-        serverHolderPredicate = s -> (s != null && s.getNumberOfSegmentsInQueue() < maxSegmentsInNodeLoadingQueue);
-      } else {
-        serverHolderPredicate = Objects::nonNull;
+      final List<ServerHolder> holders = getFilteredHolders(
+          tier,
+          params.getDruidCluster(),
+          createLoadQueueSizeLimitingPredicate(params)
+      );
+      if (holders.isEmpty()) {
+        continue;
       }
 
-      final List<ServerHolder> serverHolderList = serverQueue.stream()
-                                                             .filter(serverHolderPredicate)
-                                                             .collect(Collectors.toList());
-
-      final BalancerStrategy strategy = params.getBalancerStrategy();
-      if (availableSegments.contains(segment)) {
-        int assignedCount = assign(
-            params.getReplicationManager(),
-            tier,
-            totalReplicantsInCluster,
-            expectedReplicantsInTier,
-            totalReplicantsInTier,
-            strategy,
-            serverHolderList,
-            segment
+      final ServerHolder candidate = params.getBalancerStrategy().findNewSegmentHomeReplicator(segment, holders);
+      if (candidate == null) {
+        log.warn(
```

Comment:
This can also mean `createLoadQueueSizeLimitingPredicate` filtered out the servers, so it doesn't require that there is insufficient server capacity, it could be that existing servers are too full in their load queue.

Reply:
:+1: